features/323-login-id/design.md

```diff
@@ -0,0 +1,109 @@
+# Background
+
+## Current Design
+
+Under current design of authentication:
+
+1. Each **User** is associated with some **Principal**.
+3. Each **Principal** is associated with one **Identity Provider** (e.g. password authentication, OAuth).
+4. Password authentication principals is represented by tuple (**Login ID Key**, **Login ID**, password hash).
+5. App developer specifies a list of white-listed login ID key.
+6. To signup using password authentication, user can provide a list of tuple (login ID key, login ID). e.g. `{ "username": "test", "email": "a@example.com" }`
+7. For password authentication principals, user can login using a tuple (login ID key, login ID). e.g. `{ "email": "a@example.com" }`
+8. Login IDs are globally unique in app, without considering associated login ID keys.
+9. Auth gear does not knows the semantics of login IDs.
+
+
+## Problem
+1. Some features requires semantics of login IDs to function.
+2. Some authentication system designs may not be implemented easily in current design.
+
+
+# Use cases
+
+Considering some practical scenarios of authentication system:
+
+## GitHub
+- Supports multiple emails per user, with email verification.
+- User can add/delete any amount of emails.
+- Only verified emails would be associated with the user in public.
+- User can allow recovery email/security alert to be sent to all verified emails or a specific verified email.
+
+
+## Facebook
+- User can login with email/phone number/username, without indicating what type of value it is.
+- User can add/delete any amount of emails, which can be used to login after verification.
+- Disposable email address is blocked.
+- All emails/phones of user is verified.
+- Email communications are personalized with user language & name
+
+
+# Design Objectives
+- Allow auth related features to understand semantics of some login IDs / user metadata
+- Support multiple login ID with same login ID key
```

Comment:
I'm not sure why this becomes an objective? Or you plan to make the Login ID key meaningful semantically? (E.g. a key called `email` meant all value inside are email addresses)

Reply:
Here is the response returned by login.

```json
{
  "result": {
    "user_id": "ddaa8775-ccd3-43a3-ac9f-4362bafc4321",
    "login_ids": {
      "email": "louischan@oursky.com"
    },
    "metadata": {},
    "last_seen_at": "2019-05-31T09:43:43.863154Z",
    "created_at": "2019-05-27T08:14:55.137184Z",
    "created_by": "ddaa8775-ccd3-43a3-ac9f-4362bafc4321",
    "updated_at": "2019-05-27T08:14:55.137184Z",
    "updated_by": "ddaa8775-ccd3-43a3-ac9f-4362bafc4321",
    "verified": false,
    "verify_info": {},
    "access_token": "redacted"
  }
}
```

We can see `login_ids` is dictionary where value is string.

Suppose the developer is doing the following

```js
signupWithLoginIDs({
  email: 'johndoe@example.com',
}, 'password');
createLoginID("email", "johndoe@mycompany.com", "another_password");
createLoginID("email", "johndoe@yourcompany.com", "secret");
```

The response should look like

```json
{
  "login_ids": {
    "email": ["johndoe@example.com", "johndoe@mycompany.com", "johndoe@yourcompany.com"]
  }
}
```

Therefore, the value should really be an array if we want to support multiple login id under the same login id key.

Reply:
@louischan-oursky I understood the implication.

However, we actually have two ways to implement the above. One way is as you mentioned above, have ONE `email` key with an array of login IDs.

Another way is, in current spec [294](https://github.com/SkygearIO/features/blob/master/features/294-login-without-specifying-login-id-key/api.md), it would be:
```
{
  "login_ids": {
    "email": "johndoe@example.com",
    "email2": "johndoe@mycompany.com", 
    "email3": "johndoe@yourcompany.com"
  }
}
```

The differences is really about what we meant by LoginIDs. In the 2nd approach, `email`, `email1`, `email2`, or whatever key name, is just a name for user to annotate each loginID whatever they want.

In the new way of making each login key could have multiple login IDs, the only benefits I saw is we make some convention, of specific key name imply semantic meaning of Login IDs, such as email, or phone no?

****

BTW, please don't forget the case a lot of time login IDs could be a combination of different fields. One example is in a doctor platform app Oursky developed, doctor and patient share the same system but different logins (such an email address can have two accounts, doctor or patient).

Long story short, when we discuss #294 , these are the approaches considered:

Let's say we have the following as loginIDs:
```
{ "role": string, "email": string }
```

1. User need to serialize JSON, so Skygear LoginIDs are always string (choosen solution), and we plan to follow up in #296 Compound LoginIDs as first class citizen
1. Support Compound LoginIDs natively in the APIs. (many APIs need to redesign, hence left for #296)
1. Do not support compound loginIDs, instead we just have a attributes called "Domain" for each user. This is less flexible but good enough for 90% of use case?
1. Support using a combination of metadata as login IDs (too complicated, not preferred)

So if we decide to use Login ID keys to imply the semantic meaning, it will affect how we could implement compound login IDs.

Reply:
Thanks @chpapa . You helped us to discover that we are encouraging the developer to encode compound login ids into JSON while our proposal cannot handle that case.

cc @kiootic